fastcrawler/test_utils.py

```diff
@@ -1,62 +1,73 @@
-        return response
+class Method(StrEnum):
+    GET = "GET"
+    POST = "POST"
+    PUT = "PUT"
+    DELETE = "DELETE"
 
 
-class TestClient(AioHttpEngine):
-    def __init__(self, server=TestServer(), headers=None, proxy=None):
-        self.server = server
+class MockEngine:
+    def __init__(
+        self,
+        cookies: list[SetCookieParam] | None = None,
+        headers: dict | None = None,
+        user_agent: str | None = None,
+        proxy: ProxySetting | None = None,
+        connection_limit: int | None = None,
+    ):
+        super().__init__(cookies, headers, user_agent, proxy, connection_limit)
 
-    def get(self, url):
-        response = self.server.get_response("GET")
-        return response
+    async def get(self, requests: list[Request]) -> dict[Url, RequestCycle]:
+        """GET HTTP Method for protocol to retrieve a list of URL."""
+        return await self.batch(requests, "GET")
 
-    def post(self, url, data):
-        response = self.server.get_response("POST", data=data)
-        return response
+    async def post(self, requests: list[Request]) -> dict[Url, RequestCycle]:
+        """POST HTTP Method for protocol to crawl a list of URL."""
+        return await self.batch(requests, "POST")
 
-    def delete(self, url):
-        response = self.server.get_response("DELETE")
-        return response
+    async def put(self, requests: list[Request]) -> dict[Url, RequestCycle]:
+        """PUT HTTP Method for protocol to crawl a list of URL."""
+        return await self.batch(requests, "PUT")
 
-    def put(self, url, data):
-        response = self.server.get_response("PUT", data=data)
-        return response
+    async def delete(self, requests: list[Request]) -> dict[Url, RequestCycle]:
+        """DELETE HTTP Method for protocol to crawl a list of URL."""
+        return await self.batch(requests, "DELETE")
 
 
-# prepare test
+class TestClient:
+    def __init__(
+        self,
+        mapped_routes: dict[tuple[Method, Request], Response],
+        mock_engine: EngineProto,
+    ) -> None:
+        self.mapped_routes = mapped_routes
+        self.mock_engine = mock_engine
+
+        self.__default_response = Response(status_code=404)
 
-server = TestServer()
-client = TestClient(server)
+    def get(self, request: Request):
+        return self.mapped_routes.get((Method.GET, request), self.__default_response)
 
-# Test GET method
-response = client.get("http://example.com")
-assert response == {"message": "GET request received"}
 
-# Test POST method
-data = {"key": "value"}
-response = client.post("http://example.com", data=data)
-assert response == {"message": "POST request received", "data": {"key": "value"}}
+# prepare test
 
-# Test DELETE method
-response = client.delete("http://example.com")
-assert response == {"message": "DELETE request received"}
+mapped_routes = {
+    (Method.GET, Request(url="/user/1")): Response(text=str({"id": 1, "username": "admin"})),
```

Comment:
That's not good, few things you missed
1. Response MAY have header
2. Response MAY have cookie
3. Response MUST have status_code
4. Response can be in form of JSON or HTML. (keep that in mind). don't use str(dict[str,Any]), load the json correctly.
5. Same stuff goes for request, they MAY have those. 
6. Be more generic, like /user/{user_id}/, then everything mapped to like user/1/ would be caught by this key. maybe you shouldn't use exactly a dictionary :)

Reply:
Overall, this is better than previous.
Also it needs some missing things that you mentioned.